File: 53.最大子序和.py
#
# @lc app=leetcode.cn id=53 lang=python
#
# [53] 最大子序和
#

# @lc code=start
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0

        res = -9999999
        tmp = 0
        for i in nums:
            if tmp <= 0:
                tmp = i
            else:
                tmp = tmp + i
            res = max(tmp, res)
        return res
# @lc code=end

File: 53.最大子序和.cpp
/*
 * @lc app=leetcode.cn id=53 lang=cpp
 *
 * [53] 最大子序和
 */

// @lc code=start
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if (nums.size() == 0) {
            return 0;
        }
        int res = nums[0];
        int tmp = nums[0];
        for (int i = 1; i < nums.size(); ++i) {
            if (tmp < 0) {
                tmp = nums[i];
            } else {
                tmp += nums[i];
            }
            res = max(res, tmp);
        }
        return res;
    }
};
// @lc code=end

